test/metabase/permissions_test.clj

```diff
@@ -0,0 +1,743 @@
+        "You don't have permissions to do that."))
+
+;; admin should be able to get metadata for all tables
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db1* :checkins))
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db1* :users))
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db1* :venues))
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db2* :checkins))
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db2* :users))
+(expect-with-test-data true (GET-table-query-metadata :crowberto *db2* :venues))
+
+;; normal user should only be able to get metadata for DB 1's tables
+(expect-with-test-data true  (GET-table-query-metadata :rasta *db1* :checkins))
+(expect-with-test-data true  (GET-table-query-metadata :rasta *db1* :users))
+(expect-with-test-data true  (GET-table-query-metadata :rasta *db1* :venues))
+(expect-with-test-data false (GET-table-query-metadata :rasta *db2* :checkins))
+(expect-with-test-data false (GET-table-query-metadata :rasta *db2* :users))
+(expect-with-test-data false (GET-table-query-metadata :rasta *db2* :venues))
+
+;; ops user should be able to get metadata for DB 1 or for venues in DB 2
+(expect-with-test-data true  (GET-table-query-metadata :lucky *db1* :checkins))
+(expect-with-test-data true  (GET-table-query-metadata :lucky *db1* :users))
+(expect-with-test-data true  (GET-table-query-metadata :lucky *db1* :venues))
+(expect-with-test-data false (GET-table-query-metadata :lucky *db2* :checkins))
+(expect-with-test-data false (GET-table-query-metadata :lucky *db2* :users))
+(expect-with-test-data true  (GET-table-query-metadata :lucky *db2* :venues))
+
+
+;;; ------------------------------------------------------------ POST /api/dataset (SQL query) ------------------------------------------------------------
+
+(defn- sql-query [username db]
+  (let [results ((test-users/user->client username) :post "dataset"
+                 {:database (u/get-id db)
+                  :type     :native
+                  :native   {:query "SELECT COUNT(*) FROM VENUES;"}})]
+    (if (string? results)
+      results
+      (or (:error results)
+          (get-in results [:data :rows])))))
+
+;; everyone should be able to ask SQL questions against DB 1
+(expect-with-test-data [[100]] (sql-query :crowberto *db1*))
+(expect-with-test-data [[100]] (sql-query :rasta *db1*))
+(expect-with-test-data [[100]] (sql-query :lucky *db1*))
+
+;; Only Admin should be able to ask SQL questions against DB 2. Error message is slightly different for Rasta & Lucky because Rasta has no permissions whatsoever for DB 2 while Lucky has partial perms
+(expect-with-test-data [[100]] (sql-query :crowberto *db2*))
+(expect-with-test-data "You don't have permissions to do that." (sql-query :rasta *db2*))
+(expect-with-test-data "You do not have permissions to run new native queries against database 'DB Two'." (sql-query :lucky *db2*))
```

Comment:
yeah, luckily these error messages are ones you'd only ever see if you tried to hit an endpoint that you wouldn't be able to hit from the frontend anyway. E.g. if you went and put `http://my.metabase.instance.com/api/card/100` in your browser.

Since these aren't really user-facing error messages they don't change often.